ruff.toml

```diff
@@ -56,6 +57,16 @@ ignore = [
 "setuptools/__init__.py" = ["E402"]
 "pkg_resources/__init__.py" = ["E402"]
 
+[lint.isort]
+combine-as-imports = true
+split-on-trailing-comma = false
+# Force Ruff/isort to always import setuptools before distutils in tests as long as distutils_hack is supported
+# This also ensures _distutils_hack is imported before distutils
+# https://github.com/pypa/setuptools/issues/4137
+section-order = ["future", "standard-library", "eager", "third-party", "first-party", "delayed", "local-folder"]
```

Comment:
I proposed we leave `"delayed"` as the last item in this array.

```suggestion
section-order = ["future", "standard-library", "eager", "third-party", "first-party", "local-folder", "delayed"]
```

The reasoning is the following:

Relative import statements like `from . import ..` seem to be considered "local-folder", but they still refer to `setuptools`. This means that `distutils` imports will come before setuptools imports, which we are trying to avoid.

For example this change in setuptools/command/bdist_wheel.py:

```python
from packaging import tags, version as _packaging_version
from wheel.metadata import pkginfo_to_metadata
from wheel.wheelfile import WheelFile

from distutils import log

from .. import Command, __version__
from .egg_info import egg_info as egg_info_cls
```

Reply:
Here's my reasoning:
For non-setuptools folders, like tools and tests, it doesn't change anything and it's closer to what a user of setuptools would have.

For setuptools itself: AFAIK, to import a submodule, the top module/package gets imported too. So `_distutils_hack` should still always run first. Unless I'm missing something:

![image](https://github.com/user-attachments/assets/90cd5464-ae12-4793-be48-49557cbafd0b)
![image](https://github.com/user-attachments/assets/e6b79d71-4920-439b-badd-5cb004cd2d30)
(overly zealous prints but I'm trying to remove assumptions from my proof)

Reply:
`_distutils_hack` will not be affected by putting `delayed` last, right? (Because it goes to its own group `eager`).

Reply:
As long as `eager` is before `first-party`, `local-folder` and `delayed` yeah

Just to be clear, swapping the order of "local-folder" and "delayed" works in both cases. I'm arguing for keeping the expectation and habit of putting relative imports last.
(prefering left side in this image)
![image](https://github.com/user-attachments/assets/b17fdc02-271d-4beb-a1ac-7801b37669bf)

Reply:
I see, thank you very much for clarifying!

We currently have a recommendation that `distutils` should be imported after `setuptools` expressed in https://github.com/pypa/setuptools/blob/a6f7484b25628ae351d4b96fa6254ada387ece10/_distutils_hack/__init__.py#L17-L24

So I think that honouring that is more important than the traditional approach.

I understand that indirectly we can verify that both approaches work. But the implicit approach might be fragile in the future[^1]. So in principle, I am in favour of directly ensuring that any import of `distutils` always come explicitly after the imports to `setuptools`[^2].

[^1]: For example to solve cycles, Python import machinery may start evaluating the parent module, then interrupt this evaluation in favour of children module, then proceed to finish the evaluation of the parent module.

[^2]: Unless we verify an unintended consequence and we need to revert.

Reply:
We can change that once `SETUPTOOLS_USE_DISTUTILS=stdlib` is fully removed.